### PR TITLE
feat: add caching

### PR DESCRIPTION
**Pull request by anbraten**

### CODE REVIEW

**Comment by anbraten** [comment]
@lukashass What do you think?

File: cache.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+	cache "github.com/victorspringer/http-cache"
+	"github.com/victorspringer/http-cache/adapter/memory"
+)
+
+func cacheHandle(h http.Handler, ttl int) http.Handler {
+
+	memcached, err := memory.NewAdapter(
+		memory.AdapterWithAlgorithm(memory.LRU),
+		memory.AdapterWithCapacity(10_000_000),
```
Comment (lukashass) [inline cache.go:17]: @anbraten What is this value? Might be good to document the unit.

Will this fit every use case? Maybe it should be configurable.

File: cache.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	cacheClient, err := cache.NewClient(
+		cache.ClientWithAdapter(memcached),
+		cache.ClientWithTTL(time.Duration(ttl)*time.Second),
+		cache.ClientWithRefreshKey("opn"),
```
Comment (lukashass) [inline cache.go:28]: @anbraten What does `opn` stand for?

**Review by lukashass — APPROVED** [review]
Looks good!
How did you test if this actually works?

File: cache.go (hunk 2 above)
Comment (anbraten) [reply cache.go:28]: @lukashass It can be used as query parameter to refresh / skip the cache.

File: cache.go (hunk 1 above)
Comment (anbraten) [reply cache.go:17]: @lukashass I took it from the documentation of `http-cache``from what I can tell from the source-code of it it is the maximum number of items which can be stored in the cache. What is not checked is the size of a single item in the cache. So using it with a large download file for example could be a problem.

Comment (lukashass) [reply cache.go:17]: @anbraten Alright I though it would be some kind of bytes. I guess since the cache is not enabled by default it should be fine.

**Comment by anbraten** [comment]
I tested it with `ab -n 1000000 -c 1000 http://localhost:8080/` in comparison to a Caddy-Server serving the same static `index.html`:

s3ww with `--use-cache --cache-ttl 3600`:
```
Document Path:          /
Document Length:        31414 bytes

Concurrency Level:      1000
Time taken for tests:   104.156 seconds
Complete requests:      1000000
Failed requests:        0
Total transferred:      31601000000 bytes
HTML transferred:       31414000000 bytes
Requests per second:    9600.96 [#/sec] (mean)
Time per request:       104.156 [ms] (mean)
Time per request:       0.104 [ms] (mean, across all concurrent requests)
Transfer rate:          296288.93 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0   28   6.2     28      57
Processing:    23   76  10.2     75     572
Waiting:        4   24   8.9     23     503
Total:         43  104   9.6    102     572
```

Caddy:
```
Document Path:          /
Document Length:        31414 bytes

Concurrency Level:      1000
Time taken for tests:   114.932 seconds
Complete requests:      1000000
Failed requests:        0
Total transferred:      31635000000 bytes
HTML transferred:       31414000000 bytes
Requests per second:    8700.80 [#/sec] (mean)
Time per request:       114.932 [ms] (mean)
Time per request:       0.115 [ms] (mean, across all concurrent requests)
Transfer rate:          268798.49 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0   29   6.5     28      72
Processing:    14   86  11.1     86     307
Waiting:        2   27  11.0     26     269
Total:         47  115   9.6    114     323
```

Using s3www without cache will throw a lot of the following errors and doesn't work:

```
2021/04/26 14:21:44 http: Accept error: accept tcp 127.0.0.1:8082: accept4: too many open files; retrying in 5ms
2021/04/26 14:21:44 http: Accept error: accept tcp 127.0.0.1:8082: accept4: too many open files; retrying in 10ms
```

**Pull request merged**